File: moderate/54_cash_register.py
import sys

AMOUNT_NAMES  = {
    1 : 'PENNY',
    5 : 'NICKEL',
    10 : 'DIME',
    25 : 'QUARTER',
    50 : 'HALF DOLLAR',
    100 : 'ONE',
    200 : 'TWO',
    500 : 'FIVE',
    1000 : 'TEN',
    2000 : 'TWENTY',
    5000 : 'FIFTY',
    10000 : 'ONE HUNDRED'
}
AMOUNTS = sorted(AMOUNT_NAMES, reverse=True)

for line in open(sys.argv[1]):
    price, tendered = line.split(';')
    change = int(round(float(tendered) * 100)) - int(round(float(price) * 100))

    if change < 0:
        print('ERROR')
    elif change == 0:
        print('ZERO')
    else:
        money = []
        for amount in AMOUNTS:
            n, change = divmod(change, amount)
            if n > 0:
                money.extend([AMOUNT_NAMES[amount]] * int(n))
        print(','.join(money))

File: moderate/54_cash_register.c
#include <stdio.h>
#include <math.h>

#define LINE_LEN    1024

int AMOUNTS[] = {10000, 5000, 2000, 1000, 500, 200, 100, 50, 25, 10, 5, 1};
char *AMOUNT_NAMES[] = {"ONE HUNDRED", "FIFTY", "TWENTY", "TEN", "FIVE",
                        "TWO", "ONE", "HALF DOLLAR", "QUARTER", "DIME",
                        "NICKEL", "PENNY"};

int main(int argc, const char * argv[]) {
    FILE *file = fopen(argv[1], "r");
    char line[LINE_LEN];

    while (fgets(line, LINE_LEN, file)) {
        float price, tendered;
        sscanf(line, "%f;%f", &price, &tendered);
        int change = (int)round(tendered * 100) - (int)round(price * 100);

        if (change < 0)
            printf("ERROR\n");
        else if (change == 0)
            printf("ZERO\n");
        else {
            int print_comma = 0;
            for (int i=0; i<sizeof(AMOUNTS)/sizeof(int); i++) {
                int n = change / AMOUNTS[i];
                change = change % AMOUNTS[i];
                for (; n>0; n--)
                    print_comma = printf("%s%s", print_comma ? "," : "",
                                            AMOUNT_NAMES[i]);
            }
            printf("\n");
        }
    }

    return 0;
}
